cli/tests/pcluster/api/controllers/test_cluster_compute_fleet_controller.py

```diff
@@ -5,41 +5,318 @@
+            ),
+        ],
+    )
+    def test_bad_request_on_unstable_stack(
+        self, mocker, client, stack_status, scheduler, request_body, expected_response
+    ):
+        mocker.patch(
+            "pcluster.aws.cfn.CfnClient.describe_stack",
+            return_value=cfn_describe_stack_mock_response(scheduler, stack_status),
+        )
+        config_mock = mocker.patch("pcluster.models.cluster.Cluster.config")
+        config_mock.scheduling.scheduler = scheduler
+        response = self._send_test_request(client, request_body=request_body)
+        with soft_assertions():
+            assert_that(response.status_code).is_equal_to(400)
+            assert_that(response.get_json()).is_equal_to(expected_response)
+
+    @pytest.mark.parametrize(
+        "current_status, target_status, expected_response",
+        [
+            (
+                "RUNNING",
+                "STOP_REQUESTED",
+                {
+                    "message": "Bad Request: Failed when stopping compute fleet due to "
+                    "a concurrent update of the status. Please retry the operation."
+                },
+            ),
+            (
+                "STOPPED",
+                "START_REQUESTED",
+                {
+                    "message": "Bad Request: Failed when starting compute fleet due to "
+                    "a concurrent update of the status. Please retry the operation."
+                },
+            ),
+        ],
+    )
+    def test_concurrent_update_bad_request(self, mocker, client, current_status, target_status, expected_response):
+        """Test when the dynamodb put_item request generates concurrent issue exception."""
+        mocker.patch(
+            "pcluster.aws.cfn.CfnClient.describe_stack",
+            return_value=cfn_describe_stack_mock_response("slurm"),
+        )
+        config_mock = mocker.patch("pcluster.models.cluster.Cluster.config")
+        config_mock.scheduling.scheduler = "slurm"
+        dynamodb_item = {"Item": {"Status": current_status}}
+        last_updated_time = str(datetime.now())
+        dynamodb_item["Item"]["Last_updated_time"] = last_updated_time
+        mocker.patch("pcluster.aws.dynamo.DynamoResource.get_item", return_value=dynamodb_item)
+        mocker.patch(
+            "pcluster.aws.dynamo.DynamoResource.put_item",
+            side_effect=AWSClientError(
+                function_name="put_item",
+                message="Conditional Check Failed message from boto3",
+                error_code=AWSClientError.ErrorCode.CONDITIONAL_CHECK_FAILED_EXCEPTION.value,
+            ),
         )
-        assert_that(response.status_code).is_equal_to(200)
+        response = self._send_test_request(client, request_body={"status": target_status})
+        with soft_assertions():
+            assert_that(response.status_code).is_equal_to(400)
+            assert_that(response.get_json()).is_equal_to(expected_response)
+
+
+class TestDescribeComputeFleetStatus:
```

Comment:
can you add a test to verify what happens when:
* dynamodb table/item does not exist
* batch CE does not exist

Reply:
Done.
The new logic is:
as long as the stack is not in stable status or UPDATE_IN_PROCESS, the status of the compute fleet will be UNKNOWN. 
The reason is: we read the AWS Batch compute environment arn from the stack output. So when the AWS Batch stack is in CREATE_IN_PROGRESS status, the get compute fleet status operation will get `None` for the arn of the compute environment. Then we got the ParamValidationError from boto3 because the parameter is set to `None`. To avoid catching all ParamValidationError and return "UNKNOWN" status, I had to check the stack is in stable status or UPDATE_IN_PROCESS

Reply:
I added new unit tests for the absence of dynamodb table/item.

For both Slurm and AWS Batch,
I added new unit tests to check UNKNOWN status is returned if stack is not in the stable status.

I didn't add unit test for AWS Batch when the stack is in stable status but the compute environment does not exist because I couldn't think of the real use case